File: Libs/SecretSharing/Messages/MultStepVerificationPoly.cs
using System.Collections.Generic;

namespace Unm.DistributedSystem.SecretSharing
{
	public class MultStepVerificationPoly : Share
	{
		internal IList<Zp> RxPolynomial_Renamed;

		public MultStepVerificationPoly()
			: base(ShareType.MULT_STEP_VERIFY_POLY)
		{
		}

		public MultStepVerificationPoly(IList<Zp> RxPolynomial)
			: base(ShareType.MULT_STEP_VERIFY_POLY)
		{
			this.RxPolynomial_Renamed = RxPolynomial;
		}

		public virtual IList<Zp> RxPolynomial
		{
			get
			{
				return RxPolynomial_Renamed;
			}
			set
			{
				this.RxPolynomial_Renamed = value;
			}
		}

		//public override byte[] writeToByteArray()
		//{
		//    var bs = new BitStream();
		//    bs.writeMessageType(MessageType.MULT_STEP_VERIFY_POLY);
		//    writeToBitStreamNoHeader(bs);
		//    bs.close();
		//    return bs.ByteArray;
		//}

		//public override void loadFromByteArrayNoHeader(BitStream bs, int prime)
		//{
		//    if (bs.readBoolean())
		//        RxPolynomial_Renamed = bs.readList(prime);
		//}

		//public override void writeToBitStreamNoHeader(BitStream bs)
		//{
		//    bs.writeBoolean(RxPolynomial_Renamed != null);
		//    if (RxPolynomial_Renamed != null)
		//        bs.writeList(RxPolynomial_Renamed);
		//}

		public override bool Equals(object obj)
		{
			if (!(obj is MultStepVerificationPoly))
				return false;
			MultStepVerificationPoly second = (MultStepVerificationPoly)obj;
			if (RxPolynomial_Renamed == null && second.RxPolynomial_Renamed != null || RxPolynomial_Renamed != null && second.RxPolynomial_Renamed == null)
				return false;
			if (RxPolynomial_Renamed == null && second.RxPolynomial_Renamed == null)
				return true;
			if (RxPolynomial_Renamed.Count != second.RxPolynomial_Renamed.Count)
				return false;
			for (int i = 0; i < RxPolynomial_Renamed.Count; i++)
			{
				if (!(RxPolynomial_Renamed[i] == second.RxPolynomial_Renamed[i]))
					return false;
			}
			return true;
		}
	}
}

File: Libs/TauMPC/src/mpc/sendables/MultStepVerificationPoly.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package mpc.sendables;

import mpc.communication.BitStream;
import mpc.finite_field_math.Zp;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;


public class MultStepVerificationPoly extends Sendable{

    List<Zp> RxPolynomial;

    public MultStepVerificationPoly() {
    }


    public MultStepVerificationPoly(List<Zp> RxPolynomial) {
        this.RxPolynomial = RxPolynomial;
    }


    public List<Zp> getRxPolynomial() {
        return RxPolynomial;
    }

    public void setRxPolynomial(List<Zp> RxPolynomial) {
        this.RxPolynomial = RxPolynomial;
    }

    @Override
    public byte[] writeToByteArray() throws IOException {
        BitStream bs = new BitStream();
        bs.writeMessageType(MessageType.MULT_STEP_VERIFY_POLY);      
        writeToBitStreamNoHeader(bs);        
        bs.close();
        return bs.getByteArray();
    }

    @Override
    protected void loadFromByteArrayNoHeader(BitStream bs, int prime) throws IOException {
        if (bs.readBoolean()) {
            RxPolynomial = bs.readList(prime);
        }
    }

    @Override
    public void writeToBitStreamNoHeader(BitStream bs) {
        bs.writeBoolean(RxPolynomial != null);
        if (RxPolynomial != null){
            bs.writeList(RxPolynomial);
        }
    }  
    
    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof MultStepVerificationPoly)){
            return false;
        }
        MultStepVerificationPoly second = (MultStepVerificationPoly)obj;
        if (RxPolynomial == null && second.RxPolynomial != null || RxPolynomial != null && second.RxPolynomial == null) {
            return false;
        }
        if (RxPolynomial == null && second.RxPolynomial == null) {
            return true;
        }
        if (RxPolynomial.size() != second.RxPolynomial.size()) {
            return false;
        }
        for (int i = 0; i < RxPolynomial.size(); i++) {
            if (!(compareSecrets(RxPolynomial.get(i), second.RxPolynomial.get(i)))) {
                return false;
            }
        }
        return true;
    }
    
    public static void main(String[] args) throws IOException{
        int prime = 7;
        List<Zp> rxPoly = new ArrayList<Zp>();
        for (int i = 0; i < 6; i++) {
            if ((i%2) == 0){
                rxPoly.add(null);
                continue;
            }
            rxPoly.add(new Zp(prime, 3 * i));
        }
        MultStepVerificationPoly multStep = new MultStepVerificationPoly(rxPoly);
        byte[] ba = multStep.writeToByteArray();
        Sendable sendable = Sendable.loadFromByteArray(ba, prime);
        boolean b = sendable.equals(multStep);
        int k = 3;
    }
}
